### ISSUE TITLE
Syntactic support for calculational proofs

### ISSUE THREAD

**Issue by nikswamy** [FStarLang/FStar#1549]
Cf. issue #1522

**Comment by nikswamy** [comment]
https://github.com/FStarLang/FStar/issues/1522#issuecomment-425988450

**Comment by parno** [comment]
The prototype @mtzguido showed off in #1522 looks cool; it's neat that you can do everything via tactics.  It still seems like it would be worthwhile including some additional syntactic support as the reversed order of `finish` and `init` is a bit odd, and the big accumulation of parentheses at the end might be painful to manage.  Syntax that desugared to something like what's in that post could work.  

Operationally, I'm not sure what it means that the results are `squash` types.  Also, one valuable aspect of Dafny's `calc` is that the SMT context for each step is "fresh", in the sense that calling a lemma to justify a given step only adds that lemma's information to the context for proving that one step; the other steps don't see that information.  I'm not familiar enough with the tactic framework to see if that's the case here.  Finally, Dafny defaults to using equality between steps, but you can also override that default, for the entire calc or on a per-step basis, to use, e.g., inequalities too (as long as you don't try something silly that doesn't chain, like a `<` step followed by a `>` step.

**Comment by mtzguido** [comment]
Yes, better syntax is needed. I should have that we spoke about it at the meeting, and were wondering about how much internal support would be needed for calc statements, and apparently not much. (If we had mixfix operators, we wouldn't need any, I think.)

I think we could use `Lemma`s instead of `squash`es, and permit other preorders (`<`, `>`, ...) easily, I can make that happen. The context cleanliness you mention should also be doable, but I have to take a better look.

Also note that this is not tactic-dependent at all. The only use of tactics in that file is in the last example, where I checked that one can actually use tactics if needed, but one doesn't at all have to. I'm moving the `open FStar.Tactics` to make that clearer :).

**Comment by mtzguido** [comment]
Made some progress in `guido_calc`. There is now new syntax copied mostly from Dafny to make usage of the `Calc` module easier. I haven't yet generalized beyond preorders and allowed to combine relations (trying to find a clean way to do that, and without bringing tactics into the mix). The following now works, but note that the `==` in the steps are all ignored, only the first one matters:
```fstar
module CalcTest

open FStar.Mul

assume val a : int
assume val b : int
assume val c : int
assume val d : int
assume val e : int

let test (h1 : unit -> Lemma (a * (c + 1) = 24))
         (h2 : unit -> Lemma (b + d == 25)) =
  calc (==) {
    a + b + a * c + d - (e - e);
   == { (* reorder, e-e = 0 *) }
    a + a * c + b + d;
   == { (* distributivity *) }
    a * (c + 1) + b + d;
   == { h1 () }
    24 + b + d;
   == { h2 () }
    24 + 25;
   == { }
    49;
  }
```

**Comment by mtzguido** [comment]
Oh, also note the semicolons to terminate each expression as in Dafny. Without them we get parser conflicts, but an alternative is to only allow "atomic" terms there, and parenthesize as needed. I think this looks better, but feedback very welcome of course.

**Comment by parno** [comment]
Nice!  That looks great to me.  @chris.hawblitzel might have additional thoughts or suggestions.

**Comment by mtzguido** [comment]
See PR #1607.

**Comment by parno** [comment]
Very exciting!  Looking forward to trying it out.

**Comment by parno** [comment]
FWIW, I tried converting the proof of `lemma_carry_prime` from my old comment-style calc (https://github.com/project-everest/hacl-star/blob/019d964b4156c3c2fd3a441d86de703cc6848820/vale/code/crypto/ecc/curve25519/FastHybrid_helpers.fsti#L37) to an actual F* calc (https://github.com/project-everest/hacl-star/blob/ac23adacc9535ec321e4baa08d1699eda2e97e64/vale/code/crypto/ecc/curve25519/FastHybrid_helpers.fsti#L31) and everything still goes through nicely.  Exciting!

**Comment by parno** [comment]
As I started working with `calc`, I realized I'm not sure what the semantics of the initial operator, e.g., the equals in `calc (==) {`.  In Dafny, it sets the default, so you can skip writing "== { ... }", but that doesn't seem to be the case here.  Does it dictate the final relation between the first value and the final value?

**Comment by mtzguido** [comment]
> Does it dictate the final relation between the first value and the final value?

Yes, exactly. A calc proof basically first builds a chain of expressions related by varying relations (i.e. the ones in each step, just before `{ ... }`), and then tries to eliminate it into the declared top-level relation (i.e. the one after `calc`). Dafny, IIUC, supports a fixed set of relations and has a graph specifying which ones compose into which others. We instead ask the SMT solver if any chain of the shape given has its endpoints related by the top-level relation.

That said, I think we could support making the top-level relation be the default, if that's desirable.

**Comment by parno** [comment]
Okay, makes sense.  A default might be nice, but certainly isn't critical.  The current implementation is already super useful!

**Comment by Chris-Hawblitzel** [comment]
I find that the repetition of `==` makes it easier to read -- you can spot the differences between the values and the proofs more easily than in Dafny, just by looking for the `==`.

**Comment by parno** [comment]
True, it does give the proof a more uniform look.

**Comment by nikswamy** [comment]
btw, @nickgian , you may be interested to see this evolution of the work you did a couple of summers ago.

**Comment by mtzguido** [comment]
@parno @Chris-Hawblitzel: how has it been going with calc proofs? Any major pain points? Wondering if I should close this issue and perhaps open more specific ones.
